runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/wire/LengthPrefixUnknownCoders.java

```diff
@@ -62,109 +55,69 @@
-      }
-      updatedCoder.addComponentCoderIds(newComponentCoderId);
-      rvalBuilder.getComponentsBuilder().putCoders(newComponentCoderId, componentCoder.getCoder());
-      // Insert all component coders of the component coder.
-      rvalBuilder.getComponentsBuilder().putAllCoders(
-          componentCoder.getComponents().getCodersMap());
+    if (newCoder.equals(currentCoder)) {
+      return coderId;
     }
-    rvalBuilder.setCoder(updatedCoder);
+    String newCoderId = generateUniqueId(coderId + "-length_prefix", components::containsCoders);
+    components.putCoders(newCoderId, newCoder);
+    return newCoderId;
+  }
 
-    return rvalBuilder.build();
+  private static Coder addForModelCoder(
+      Coder coder, RunnerApi.Components.Builder components, boolean replaceWithByteArrayCoder) {
+    RunnerApi.Coder.Builder builder = coder.toBuilder().clearComponentCoderIds();
+    for (String componentCoderId : coder.getComponentCoderIdsList()) {
+      builder.addComponentCoderIds(
+          addLengthPrefixedCoder(componentCoderId, components, replaceWithByteArrayCoder));
+    }
+    return builder.build();
   }
 
   // If we are handling an unknown URN then we need to wrap it with a length prefix coder.
   // If requested we also replace the unknown coder with a byte array coder.
-  private static MessageWithComponents lengthPrefixUnknownCoder(
-      String coderId,
-      RunnerApi.Components components,
-      boolean replaceWithByteArrayCoder) {
-    MessageWithComponents.Builder rvalBuilder = MessageWithComponents.newBuilder();
-    RunnerApi.Coder currentCoder = components.getCodersOrThrow(coderId);
-
-    String lengthPrefixComponentCoderId = coderId;
-    if (replaceWithByteArrayCoder) {
-      return createLengthPrefixByteArrayCoder(coderId, components);
-    } else {
-      rvalBuilder.getComponentsBuilder().putCoders(coderId, currentCoder);
-    }
-
-    rvalBuilder
-        .getCoderBuilder()
-        .addComponentCoderIds(lengthPrefixComponentCoderId)
+  private static Coder wrapWithLengthPrefixCoder(String coderId) {
+    Coder.Builder lengthPrefixed = Coder.newBuilder().addComponentCoderIds(coderId);
+    lengthPrefixed
         .getSpecBuilder()
         .getSpecBuilder()
-        .setUrn(ModelCoders.LENGTH_PREFIX_CODER_URN);
-    return rvalBuilder.build();
+        .setUrn(ModelCoders.LENGTH_PREFIX_CODER_URN)
+        .build();
+    return lengthPrefixed.build();
   }
 
-  private static MessageWithComponents createLengthPrefixByteArrayCoder(
-      String coderId,
-      RunnerApi.Components components) {
-    MessageWithComponents.Builder rvalBuilder = MessageWithComponents.newBuilder();
-
-    String byteArrayCoderId = generateUniqueId(
-        coderId + "-byte_array",
-        Sets.union(components.getCodersMap().keySet(),
-            rvalBuilder.getComponents().getCodersMap().keySet()));
+  /** Adds the (singleton) length-prefixed byte array coder. */
```

Comment:
This will get added multiple times so its not really a singleton

Reply:
Please fix the comment since I still think this applies.

Reply:
Not anymore - I changed the code to not add coders that are already in the map.